File: LeetCode/20.Valid_Parentheses/Mario/20.Valid_Parentheses.py
#!/usr/bin/env python3


class Solution:
    def isValid(self, s: str) -> bool:
        dic = {")": "(", "]": "[", "}": "{"}
        stack = []

        for c in s:
            if stack and c in dic and dic[c] == stack[-1]:
                stack.pop()
            else:
                stack.append(c)

        return len(stack) == 0

File: LeetCode/20.Valid_Parentheses/Mario/20.Valid_Parentheses.swift
class Solution {
    func isValid(_ s: String) -> Bool {
        let brackets: [Character: Character] = [")": "(", "]": "[", "}": "{"]
        var stack: [Character] = []

        for ch in s {
            if let need = brackets[ch] {
                guard let top = stack.popLast(), top == need else {
                    return false
                }
            } else {
                stack.append(ch)
            }
        }

        return stack.isEmpty
    }
}
